Practice_02-display/user_space_test_program/main.c

```diff
@@ -0,0 +1,201 @@
+	printf("smem_start  : 0x%08lX\n", ffsi->smem_start);
+	printf("smem_len    : %u\n", ffsi->smem_len);
+	printf("line_length : %u\n", ffsi->line_length);
+	printf("=================================\n");
+}
+
+void plot_pixel(struct fb_config *fb, int x, int y, short color)
+{
+	int offset = (y * fb->width) + x;
+	*((fb->data) + offset) = color;
+}
+
+void draw_rect(struct fb_config *fb, int x, int y, int w, int h, short color)
+{
+	int dx, dy;
+
+	for (dy = 0; dy < h; dy++) {
+		for (dx = 0; dx < w; dx++) {
+			plot_pixel(fb, x + dx, y + dy, color);
+		}
+	}
+}
+
+void clear_screen(struct fb_config *fb, short color)
+{
+	draw_rect(fb, 0, 0, fb->width, fb->height, color);
+}
+
+void usage(const char *argv_0)
+{
+	printf("\nUsage %s: [-c<color>] [-B<border>] [-i<buffer index>]\n", argv_0);
+	printf("  Color default to 0x0000\n");
+	printf("  The border color applies to all rgb and is 10 pixels wide\n");
+	printf("  If border is not provided, none is drawn.\n");
+	exit(1);
+}
+
+int main(int argc, char **argv)
+{
+	int fd;
+	unsigned long offset;
+	struct fb_var_screeninfo fvsi;
+	struct fb_fix_screeninfo ffsi;
+	struct fb_config fb;
+	int border = -1;
+	int opt;
+	int index = 0;
+	int loop = 0;
+	short color = 0;
+
+	while ((opt = getopt(argc, argv, "c:B:i:h")) != -1) {
+		switch (opt) {
+		case 'c':
+			color = strtol(optarg, NULL, 0);
+			break;
+
+		case 'B':
+			border = 0x1F & strtol(optarg, NULL, 0);
+			break;
+
+		default:
+			usage(argv[0]);
+			break;
+		}
+	}
+
+	memset(&fb, 0, sizeof(fb));
+
+	if ((fd = open("/dev/fb0", O_RDWR)) < 0) {
+		perror("open");
+		exit(1);
+	}
+
+	if (ioctl(fd, FBIOGET_VSCREENINFO, &fvsi) < 0) {
+		perror("ioctl(FBIOGET_VSCREENINFO)");
+		close(fd);
+		exit(1);
+	}
+
+	dump_vscreeninfo(&fvsi);
+
+	if (ioctl(fd, FBIOGET_FSCREENINFO, &ffsi) < 0) {
+		perror("ioctl(FBIOGET_FSCREENINFO)");
+		close(fd);
+		exit(1);
+	}
+
+	dump_fscreeninfo(&ffsi);
+
+	fb.fd = fd;
+	fb.width = fvsi.xres;
+	fb.height = fvsi.yres;
+	fb.height_virtual = fvsi.yres_virtual;
+	fb.bpp = fvsi.bits_per_pixel;
+	fb.stride = ffsi.line_length;
+	fb.red_offset = fvsi.red.offset;
+	fb.red_length = fvsi.red.length;
+	fb.green_offset = fvsi.green.offset;
+	fb.green_length = fvsi.green.length;
+	fb.blue_offset = fvsi.blue.offset;
+	fb.blue_length = fvsi.blue.length;
+	fb.transp_offset = fvsi.transp.offset;
+	fb.transp_length = fvsi.transp.length;
+	fb.buffer_num = fb.height_virtual / fb.height;
+	
+	short  buf_test[fb.width * fb.height];
+
+	memset(buf_test, 0x00, sizeof(buf_test));
+	fb.data = buf_test;
+
+	while (1) {
+		if (border == -1) {
+			clear_screen(&fb, color);
+			break;
+		} else {
+			clear_screen(&fb, 0x0000);
+
+			draw_rect(&fb, border, border, fb.width - (border * 2 ),
+				fb.height - (border * 2), rand()&0xFFFF);
+
+			border += 5;
+				if (border >= 50) border -= 50 ;
+		}
+
+	lseek(fd, 0, SEEK_SET);	
+	write(fd, (char*)fb.data, 2*sizeof(buf_test));
+	usleep(100000);
+	}
+
+	close(fd);
+
+	return 0;
+}
+
```

Comment:
This program is nice. But it is possible to make it better and free of warnings.
gcc -O2 -Wall -c -o main.o main.c
main.c: In function ‘main’:
main.c:115:6: warning: unused variable ‘loop’ [-Wunused-variable]
  int loop = 0;
      ^
main.c:114:6: warning: unused variable ‘index’ [-Wunused-variable]
  int index = 0;
      ^
main.c:108:16: warning: unused variable ‘offset’ [-Wunused-variable]
  unsigned long offset;
                ^
main.c:193:2: warning: ignoring return value of ‘write’, declared with attribute warn_unused_result [-Wunused-result]
  write(fd, (char*)fb.data, 2*sizeof(buf_test));
  ^

Reply:
The warnings have been fixed